src/chainstate/stacks/boot/lockup.clar

```diff
@@ -0,0 +1,148 @@
+(define-map owner-unlock-heights-len
+    ((owner principal))
+    ((len uint))
+)
+
+;; Total number of additional tokens unlocked at a particular Stacks block height.
+;; Used to calculate the number of liquid STX at each block, which is needed for
+;; PoX.
+(define-map unlocked-stx-per-block
+    ((stx-height uint))
+    ((total-unlocked uint))
+)
+
+;; Unlock someone's STX for this block.
+;; This is a private method (not callable from smart contracts), but is called by the
+;; miner and by the block-processing code in the Stacks node directly to ensure that purchased
+;; tokens unlock at the right block heights.  The miner / block-processor uses the
+;; locked-stx-len map to determine the possible values for index, and iterates over
+;; the rows in locked-stx to process each token owner's token unlocks.
+;; If this method fails -- i.e. returns (err ...), then the Stacks node aborts.
+(define-private (internal-unlock-stx (index uint))
+    (let (
+        ;; If there's no unlock data, then it's a fatal node error.
+        ;; No one's tokens should be unlocked twice, and the node shouldn't ever ask
+        ;; to unlock tokens that don't exist.
+        (unlock-data
+            (unwrap!
+                (map-get? internal-locked-stx { stx-height: block-height, index: index })
+                (err ERR-UNLOCK-UNREACHABLE)))
+
+        (this-contract (as-contract tx-sender))
+    )
+    (begin
+        (unwrap!
+            (stx-transfer? (get unlock-ustx unlock-data) this-contract (get owner unlock-data))
+            (err ERR-UNLOCK-UNREACHABLE))
+        
+        ;; never process this row again
+        (map-delete internal-locked-stx { stx-height: block-height, index: index })
+        (ok true)
+    ))
+)
+
+;; Determine how many principals need to have their tokens unlocked at this block height.
+;; This is only called by the miner / node internally, so it can call (internal-unlock-stx).
+(define-private (internal-get-num-unlocks)
+    (default-to
+        u0
+        (get len (map-get? internal-locked-stx-len { stx-height: block-height })))
+)
+
+;; Schedule a token unlock record at a particular Stacks block height, for a particular owner.
+;; Used only during the Stacks 1.0 import to populate this contract's tables.
+;; NOTE: should be called in ascending order in block height
+(define-private (schedule-token-unlock (owner principal) (stx-height uint) (unlock-ustx uint) (metadata (buff 32)))
+    (let (
+        (cur-stx-unlock-len
+            (default-to
+                u0
+                (get len (map-get? internal-locked-stx-len { stx-height: stx-height }))))
+
+        (cur-owner-unlock-len
+            (default-to
+                u0
+                (get len (map-get? owner-unlock-heights-len { owner: owner } ))))
+
+        (cur-stx-unlocked
+            (default-to
+                u0
+                (get total-unlocked (map-get? unlocked-stx-per-block { stx-height: stx-height }))))
+    )
+    (begin
```

Comment:
The `begin` here is not required (it's implicit after `(let <bindings> <statements...>)`)

Reply:
The dangling close-parens on line 123 is misplaced. It does not match the open-parens before the `let` symbol but rather the open-parens afterwards, before the bindings.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/7c8a0b682fa4794cf05355a1434545097100a351